doc/website/blog/2023-02-23-simple-cms-with-ent.mdx

```diff
@@ -0,0 +1,779 @@
+	author, err := s.client.User.Query().Only(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if err := s.client.Post.Create().
+		SetTitle(r.FormValue("title")).
+		SetBody(r.FormValue("body")).
+		SetAuthor(author).
+		Exec(r.Context()); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	http.Redirect(w, r, "/", http.StatusFound)
+}
+```
+As you can see, the handler currently loads the *only* user from the `users` table (since we have yet to create a user management system or login capabilities). `Only` will fail unless exactly one result is retrieved from the database.
+
+Next, our handler creates a new post, by setting the title and body fields to values retrieved from `r.FormValue`. This is where Go stores all of the form input passed to an HTTP request.
+
+After creating the handler, we should wire it to our router:
+```go
+// newRouter creates a new router with the blog handlers mounted.
+func newRouter(srv *server) chi.Router {
+	r := chi.NewRouter()
+	r.Use(middleware.Logger)
+	r.Use(middleware.Recoverer)
+	r.Get("/", srv.index)
+	// highlight-next-line
+	r.Post("/add", srv.add)
+	return r
+}
+```
+Next, we can add an HTML `<form>` component that will be used by our user to write their content:
+```html
+<div class="col-md-12">
+	<hr/>
+	<h2>Create a new post</h2>
+	<form action="/add" method="post">
+		<div class="mb-3">
+			<label for="title" class="form-label">Title</label>
+			<input name="title" type="text" class="form-control" id="title" placeholder="Once upon a time..">
+		</div>
+		<div class="mb-3">
+			<label for="body" class="form-label">Body</label>
+			<textarea name="body" class="form-control" id="body" rows="8"></textarea>
+		</div>
+		<div class="mb-3">
+			<button type="submit" class="btn btn-primary mb-3">Post</button>
+		</div>
+	</form>
+</div>
+```
+
+Also, let's add a nice touch, where we display the blog posts from new to old. To do this, modify the `index` handler to order the posts in a descending order using the `created_at` column:
+```go
+posts, err := s.client.Post.
+	Query().
+	WithAuthor().
+	// highlight-next-line
+	Order(ent.Desc(post.FieldCreatedAt)).
+	All(ctx)
+```
+
+Finally, let's add another unit test that verifies the add post flow works as expected:
+```go
+func TestAdd(t *testing.T) {
+	client := enttest.Open(t, "sqlite3", "file:ent?mode=memory&cache=shared&_fk=1")
+	defer client.Close()
+	err := seed(context.Background(), client)
+	require.NoError(t, err)
+
+	srv := newServer(client)
+	r := newRouter(srv)
+
+	ts := httptest.NewServer(r)
+	defer ts.Close()
+
+	// Post the form.
+	resp, err := ts.Client().PostForm(ts.URL+"/add", map[string][]string{
+		"title": {"Testing, one, two."},
+		"body":  {"This is a test"},
+	})
+	require.NoError(t, err)
+	// We should be redirected to the index page and receive 200 OK.
+	require.Equal(t, http.StatusOK, resp.StatusCode)
```

Comment:
Will this really be green? You return a 302 form the handler, right?

Reply:
Test passes. Yes